tutorials/math/combinatorics.md

```diff
@@ -1,10 +1,244 @@
+
+Notice that we are not required to return the value after taking modulo and the constraints allow for a $O(n^2)$ precomputation. Thus, we will
+simply construct the entire Pascal's Triangle and query it everytime to calculate the answer.
+
+<Tabs>
+<TabItem value="cpp" label="C++">
+
+```cpp
+class Solution {
+public:
+    int uniquePaths(int m, int n) {
+        // The upper limit is m + n - 2 = 198
+        vector<vector<long long int>> PascalTriangle(199, vector<long long int> ());
+        PascalTriangle[0] = {1};
+        // Calculating every row of the triangle
+        for (int i = 1; i <= 198; i++) {
+            PascalTriangle[i].push_back(1);
+            // Using the recurrence relation.
+            for (int j = 1; j < i; j++) {
+                // Take mod with INT_MAX as otherwise some values may overflow.
+                PascalTriangle[i].push_back((PascalTriangle[i - 1][j] + PascalTriangle[i - 1][j - 1]) % INT_MAX);
+            }
+            PascalTriangle[i].push_back(1);
+        }
+        // Query the final answer
+        return PascalTriangle[m + n - 2][n - 1];
+    }
+};
+```
+
+</TabItem>
+</Tabs>
+
+NOTE: Since every testcase only asks us to find $n \choose r$ for particular values of $n$ and $r$, we can instead of precomputing the entire
+Pascal's Triangle, just compute the paricular value of $n \choose r$ using the recurrence relation and memoization. This will lead to less time
+and space complexity, as we only calculate the values we need. Also, then we no longer need to take modulo with INT_MAX as all the values will
+fit in the "int" type as mentioned in the question.
+
+You can check the complete solution for this problem [here](../../solutions/0000-0099/unique-paths-medium)
+
+### Example #4 [2400 - Number of Ways to Reach a Position After Exactly k Steps](https://leetcode.com/problems/number-of-ways-to-reach-a-position-after-exactly-k-steps/)
+
+Let's represent going left as $-1$ and going right as $+1$. Thus, following the same idea as before, we have $k$ blanks to fill with $+1$ and
+$-1$ such that there sum is equal to $startPos - endPos$.
+
+Here we can immediately see that such will be impossible in only 2 cases:
+- The parity of $k$ and $startPos - endPos$ is different.
+- The magnitude of $k$ is less than magnitude of $startPos - endPos$.
+
+After checking for above 2 cases, we know for sure that there exists a solution. Now we can just find the number of $1's$ and $-1's$ required
+to sum to $startPos - endPos$. Let them be $a$ and $b$ respectively. Then the solution is $startPos - endPos \choose a$ $=$ $startPos - endPos 
+\choose b$
+
+If we assume that $endPos >= startPos$, then number of $-1$ is $(k - (endPos - startPos))/2$. If $endPos < startPos$, then we can count the number of $+1$, which will be the same. 
+
+Thus, we need to find $nCr$$(k, \frac{n - endPos + startPos}{2})$.
```

Comment:
The formula is not correct.

 assuming $endPos >= startPos$, then we know
```
 left + right = k -- (1)
 right - left = endPos - startPos -- (2)
 (1) + (2)
 left + right - left + right = k + endPos - startPos
 2 * right = endPos - startPos + k
 right = (endPos - startPos + k) / 2
```
 in other word, left would be k - (endPos - startPos + k) / 2)

so either $nCr$ $(k, k -  \frac{endPos - startPos + k}{2})$ or $nCr$ $(k, \frac{endPos - startPos + k}{2})$

I think you can include this proof also.

Reply:
Yeah, that's a typo on my part, not sure how I missed that